{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"deploy_model.ipynb","provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyPfysjcFFQzgP7bvb0uwZEj"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"V9o2LHWcpuEy","executionInfo":{"status":"ok","timestamp":1623139693417,"user_tz":-540,"elapsed":8246,"user":{"displayName":"code clean","photoUrl":"","userId":"00477991541347387430"}},"outputId":"d00da6e7-fee6-40fd-d8bf-b255b811f545"},"source":["# 직접! 작성한 코드를 깃허브에 올려서 가져오는 코드입니다.\n","!git clone https://github.com/piaochung/food_image_classification_and_deploy.git"],"execution_count":3,"outputs":[{"output_type":"stream","text":["Cloning into 'food_image_classification_and_deploy'...\n","remote: Enumerating objects: 39, done.\u001b[K\n","remote: Counting objects:   9% (1/11)\u001b[K\rremote: Counting objects:  18% (2/11)\u001b[K\rremote: Counting objects:  27% (3/11)\u001b[K\rremote: Counting objects:  36% (4/11)\u001b[K\rremote: Counting objects:  45% (5/11)\u001b[K\rremote: Counting objects:  54% (6/11)\u001b[K\rremote: Counting objects:  63% (7/11)\u001b[K\rremote: Counting objects:  72% (8/11)\u001b[K\rremote: Counting objects:  81% (9/11)\u001b[K\rremote: Counting objects:  90% (10/11)\u001b[K\rremote: Counting objects: 100% (11/11)\u001b[K\rremote: Counting objects: 100% (11/11), done.\u001b[K\n","remote: Compressing objects: 100% (8/8), done.\u001b[K\n","remote: Total 39 (delta 8), reused 3 (delta 3), pack-reused 28\u001b[K\n","Unpacking objects: 100% (39/39), done.\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"kvNBJ0h0p-Lf","executionInfo":{"status":"ok","timestamp":1623139693417,"user_tz":-540,"elapsed":5,"user":{"displayName":"code clean","photoUrl":"","userId":"00477991541347387430"}},"outputId":"c43e1cbf-7ce6-4676-ec57-29696e16e226"},"source":["%cd food_image_classification_and_deploy/deploy"],"execution_count":4,"outputs":[{"output_type":"stream","text":["/content/food_image_classification_and_deploy/deploy/food_image_classification_and_deploy/deploy\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"SMr673aYqFtV","executionInfo":{"status":"ok","timestamp":1623139695245,"user_tz":-540,"elapsed":1832,"user":{"displayName":"code clean","photoUrl":"","userId":"00477991541347387430"}}},"source":["import tensorflow as tf\n","import numpy as np\n","import urllib.request\n","from keras.models import load_model\n","from keras.preprocessing import image\n","from keras.applications import imagenet_utils"],"execution_count":5,"outputs":[]},{"cell_type":"code","metadata":{"id":"m0Hwof3QqK7H","executionInfo":{"status":"ok","timestamp":1623139695249,"user_tz":-540,"elapsed":10,"user":{"displayName":"code clean","photoUrl":"","userId":"00477991541347387430"}}},"source":["def save_and_get_img_path(url):\n","  url_path = url\n","  file_name = url.split('/')[-1]\n","  urllib.request.urlretrieve(url_path, filename=file_name)\n","  return file_name"],"execution_count":6,"outputs":[]},{"cell_type":"code","metadata":{"id":"Vvt_Ik_CqLqx","executionInfo":{"status":"ok","timestamp":1623139695249,"user_tz":-540,"elapsed":9,"user":{"displayName":"code clean","photoUrl":"","userId":"00477991541347387430"}}},"source":["def prepare_image(img_path):\n","  img = image.load_img(img_path, target_size=(img_width, img_height))\n","  img_array = image.img_to_array(img)\n","  img_array_expand_dims = np.expand_dims(img_array, axis=0)\n","  return tf.keras.applications.mobilenet.preprocess_input(img_array_expand_dims)"],"execution_count":7,"outputs":[]},{"cell_type":"code","metadata":{"id":"7HtNwzT3qM4H","executionInfo":{"status":"ok","timestamp":1623139695250,"user_tz":-540,"elapsed":9,"user":{"displayName":"code clean","photoUrl":"","userId":"00477991541347387430"}}},"source":["from googleapiclient.discovery import build\n","from googleapiclient.errors import HttpError\n","from oauth2client.tools import argparser"],"execution_count":8,"outputs":[]},{"cell_type":"code","metadata":{"id":"MFjiaz-CqNdn","executionInfo":{"status":"ok","timestamp":1623139695250,"user_tz":-540,"elapsed":9,"user":{"displayName":"code clean","photoUrl":"","userId":"00477991541347387430"}}},"source":["# developer_key = \"'AIzaSyCqEZTPQlS48rUKlRxe2sMvV1UXpK4F3ZM'\"\n","def build_youtube_search(developer_key):\n","  DEVELOPER_KEY = developer_key\n","  YOUTUBE_API_SERVICE_NAME=\"youtube\"\n","  YOUTUBE_API_VERSION=\"v3\"\n","  return build(YOUTUBE_API_SERVICE_NAME,YOUTUBE_API_VERSION,developerKey=DEVELOPER_KEY)"],"execution_count":9,"outputs":[]},{"cell_type":"code","metadata":{"id":"UhYWovFEqOTu","executionInfo":{"status":"ok","timestamp":1623139695250,"user_tz":-540,"elapsed":7,"user":{"displayName":"code clean","photoUrl":"","userId":"00477991541347387430"}}},"source":["def get_search_response(youtube, query):\n","  search_response = youtube.search().list(\n","    q = query,\n","    order = \"relevance\",\n","    part = \"snippet\",\n","    maxResults = 10\n","    ).execute()\n","  return search_response"],"execution_count":10,"outputs":[]},{"cell_type":"code","metadata":{"id":"oQbk2-RQqPGf","executionInfo":{"status":"ok","timestamp":1623139695251,"user_tz":-540,"elapsed":8,"user":{"displayName":"code clean","photoUrl":"","userId":"00477991541347387430"}}},"source":["def get_video_info(search_response):\n","  result_json = {}\n","  idx =0\n","  for item in search_response['items']:\n","    if item['id']['kind'] == 'youtube#video':\n","      result_json[idx] = info_to_dict(item['id']['videoId'], item['snippet']['title'], item['snippet']['description'], item['snippet']['thumbnails']['medium']['url'])\n","      idx += 1\n","  return result_json"],"execution_count":11,"outputs":[]},{"cell_type":"code","metadata":{"id":"d65kQ7ntqPoS","executionInfo":{"status":"ok","timestamp":1623139695251,"user_tz":-540,"elapsed":8,"user":{"displayName":"code clean","photoUrl":"","userId":"00477991541347387430"}}},"source":["def info_to_dict(videoId, title, description, url):\n","  result = {\n","      \"videoId\": videoId,\n","      \"title\": title,\n","      \"description\": description,\n","      \"url\": url\n","  }\n","  return result"],"execution_count":12,"outputs":[]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"cRAJ_ffhqQcc","executionInfo":{"status":"ok","timestamp":1623139698733,"user_tz":-540,"elapsed":3489,"user":{"displayName":"code clean","photoUrl":"","userId":"00477991541347387430"}},"outputId":"d51b488f-e7b9-4f19-9d6b-40557a19eaa5"},"source":["!pip install flask_ngrok"],"execution_count":13,"outputs":[{"output_type":"stream","text":["Collecting flask_ngrok\n","  Downloading https://files.pythonhosted.org/packages/af/6c/f54cb686ad1129e27d125d182f90f52b32f284e6c8df58c1bae54fa1adbc/flask_ngrok-0.0.25-py3-none-any.whl\n","Requirement already satisfied: Flask>=0.8 in /usr/local/lib/python3.7/dist-packages (from flask_ngrok) (1.1.4)\n","Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (from flask_ngrok) (2.23.0)\n","Requirement already satisfied: itsdangerous<2.0,>=0.24 in /usr/local/lib/python3.7/dist-packages (from Flask>=0.8->flask_ngrok) (1.1.0)\n","Requirement already satisfied: Jinja2<3.0,>=2.10.1 in /usr/local/lib/python3.7/dist-packages (from Flask>=0.8->flask_ngrok) (2.11.3)\n","Requirement already satisfied: click<8.0,>=5.1 in /usr/local/lib/python3.7/dist-packages (from Flask>=0.8->flask_ngrok) (7.1.2)\n","Requirement already satisfied: Werkzeug<2.0,>=0.15 in /usr/local/lib/python3.7/dist-packages (from Flask>=0.8->flask_ngrok) (1.0.1)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->flask_ngrok) (2020.12.5)\n","Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->flask_ngrok) (2.10)\n","Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests->flask_ngrok) (3.0.4)\n","Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->flask_ngrok) (1.24.3)\n","Requirement already satisfied: MarkupSafe>=0.23 in /usr/local/lib/python3.7/dist-packages (from Jinja2<3.0,>=2.10.1->Flask>=0.8->flask_ngrok) (2.0.1)\n","Installing collected packages: flask-ngrok\n","Successfully installed flask-ngrok-0.0.25\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"hVY0669WqRL5","executionInfo":{"status":"ok","timestamp":1623139698734,"user_tz":-540,"elapsed":4,"user":{"displayName":"code clean","photoUrl":"","userId":"00477991541347387430"}}},"source":["from flask import Flask, request, Response\n","from flask_ngrok import run_with_ngrok\n","import json"],"execution_count":14,"outputs":[]},{"cell_type":"code","metadata":{"id":"PCwNfLqZqSge","executionInfo":{"status":"ok","timestamp":1623139701000,"user_tz":-540,"elapsed":2269,"user":{"displayName":"code clean","photoUrl":"","userId":"00477991541347387430"}}},"source":["app = Flask(__name__)\n","img_width, img_height = 128, 128\n","\n","# developer_key = \"\"\n","model = load_model('best_model.h5')\n","\n","class_names = ['바베큐 폭림',\n"," '비빔밥',\n"," '샌드위치',\n"," '구운 연어',\n"," '핫도그',\n"," '스테이크']\n","\n","@app.route('/')\n","def hello_world():\n","  return \"just started\"\n","\n","@app.route('/test')\n","def dsdsd():\n","  return request.args['url']\n","\n","@app.route('/predict', methods=['GET', 'POST'])\n","def predict_image():\n","  url = request.args['url']\n","  img_path = save_and_get_img_path(url)\n","  preprocess_image = prepare_image(img_path)\n","  predictions = model.predict(preprocess_image)\n","  return json.dumps(class_names[np.argmax(predictions)], ensure_ascii=False)\n","\n","@app.route('/video_list', methods=['GET', 'POST'])\n","def get_video_list():\n","  developer_key =  request.args['key']\n","  query = request.args['query']\n","  youtube = build_youtube_search(developer_key)\n","  search_response = get_search_response(youtube, query)\n","  res = json.dumps(get_video_info(search_response), ensure_ascii=False)\n","  return Response(res, content_type='application/json; charset=utf-8')"],"execution_count":15,"outputs":[]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"BJJZ4IheqTcP","outputId":"0e15707d-97d1-4d48-a9fe-99a40410b2d6"},"source":["run_with_ngrok(app)\n","app.run()"],"execution_count":null,"outputs":[{"output_type":"stream","text":[" * Serving Flask app \"__main__\" (lazy loading)\n"," * Environment: production\n","\u001b[31m   WARNING: This is a development server. Do not use it in a production deployment.\u001b[0m\n","\u001b[2m   Use a production WSGI server instead.\u001b[0m\n"," * Debug mode: off\n"],"name":"stdout"},{"output_type":"stream","text":[" * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)\n"],"name":"stderr"},{"output_type":"stream","text":[" * Running on http://57d3797e9153.ngrok.io\n"," * Traffic stats available on http://127.0.0.1:4040\n"],"name":"stdout"},{"output_type":"stream","text":["127.0.0.1 - - [08/Jun/2021 08:08:25] \"\u001b[37mGET / HTTP/1.1\u001b[0m\" 200 -\n","127.0.0.1 - - [08/Jun/2021 08:08:25] \"\u001b[33mGET /favicon.ico HTTP/1.1\u001b[0m\" 404 -\n","127.0.0.1 - - [08/Jun/2021 08:08:41] \"\u001b[37mGET /video_list?key=AIzaSyCqEZTPQlS48rUKlRxe2sMvV1UXpK4F3ZM&&query=운동 HTTP/1.1\u001b[0m\" 200 -\n","127.0.0.1 - - [08/Jun/2021 08:27:30] \"\u001b[37mGET / HTTP/1.1\u001b[0m\" 200 -\n","127.0.0.1 - - [08/Jun/2021 08:27:31] \"\u001b[37mGET / HTTP/1.1\u001b[0m\" 200 -\n","127.0.0.1 - - [08/Jun/2021 08:27:33] \"\u001b[37mGET / HTTP/1.1\u001b[0m\" 200 -\n","[2021-06-08 08:32:27,566] ERROR in app: Exception on /predict [GET]\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 2447, in wsgi_app\n","    response = self.full_dispatch_request()\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1952, in full_dispatch_request\n","    rv = self.handle_user_exception(e)\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1821, in handle_user_exception\n","    reraise(exc_type, exc_value, tb)\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/_compat.py\", line 39, in reraise\n","    raise value\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1950, in full_dispatch_request\n","    rv = self.dispatch_request()\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1936, in dispatch_request\n","    return self.view_functions[rule.endpoint](**req.view_args)\n","  File \"<ipython-input-15-6674ec62ed14>\", line 25, in predict_image\n","    img_path = save_and_get_img_path(url)\n","  File \"<ipython-input-6-b6e971d27ede>\", line 4, in save_and_get_img_path\n","    urllib.request.urlretrieve(url_path, filename=file_name)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 247, in urlretrieve\n","    with contextlib.closing(urlopen(url, data)) as fp:\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 222, in urlopen\n","    return opener.open(url, data, timeout)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 525, in open\n","    response = self._open(req, data)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 543, in _open\n","    '_open', req)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 503, in _call_chain\n","    result = func(*args)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 1378, in http_open\n","    return self.do_open(http.client.HTTPConnection, req)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 1350, in do_open\n","    encode_chunked=req.has_header('Transfer-encoding'))\n","  File \"/usr/lib/python3.7/http/client.py\", line 1277, in request\n","    self._send_request(method, url, body, headers, encode_chunked)\n","  File \"/usr/lib/python3.7/http/client.py\", line 1288, in _send_request\n","    self.putrequest(method, url, **skips)\n","  File \"/usr/lib/python3.7/http/client.py\", line 1126, in putrequest\n","    self._output(self._encode_request(request))\n","  File \"/usr/lib/python3.7/http/client.py\", line 1206, in _encode_request\n","    return request.encode('ascii')\n","UnicodeEncodeError: 'ascii' codec can't encode characters in position 19-20: ordinal not in range(128)\n","127.0.0.1 - - [08/Jun/2021 08:32:27] \"\u001b[35m\u001b[1mGET /predict?url=http://ec2-52-79-44-86.ap-northeast-2.compute.amazonaws.com/upload/images/하이/1623141147.jpeg HTTP/1.1\u001b[0m\" 500 -\n","[2021-06-08 08:32:33,097] ERROR in app: Exception on /predict [GET]\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 2447, in wsgi_app\n","    response = self.full_dispatch_request()\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1952, in full_dispatch_request\n","    rv = self.handle_user_exception(e)\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1821, in handle_user_exception\n","    reraise(exc_type, exc_value, tb)\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/_compat.py\", line 39, in reraise\n","    raise value\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1950, in full_dispatch_request\n","    rv = self.dispatch_request()\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1936, in dispatch_request\n","    return self.view_functions[rule.endpoint](**req.view_args)\n","  File \"<ipython-input-15-6674ec62ed14>\", line 25, in predict_image\n","    img_path = save_and_get_img_path(url)\n","  File \"<ipython-input-6-b6e971d27ede>\", line 4, in save_and_get_img_path\n","    urllib.request.urlretrieve(url_path, filename=file_name)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 247, in urlretrieve\n","    with contextlib.closing(urlopen(url, data)) as fp:\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 222, in urlopen\n","    return opener.open(url, data, timeout)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 525, in open\n","    response = self._open(req, data)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 543, in _open\n","    '_open', req)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 503, in _call_chain\n","    result = func(*args)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 1378, in http_open\n","    return self.do_open(http.client.HTTPConnection, req)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 1350, in do_open\n","    encode_chunked=req.has_header('Transfer-encoding'))\n","  File \"/usr/lib/python3.7/http/client.py\", line 1277, in request\n","    self._send_request(method, url, body, headers, encode_chunked)\n","  File \"/usr/lib/python3.7/http/client.py\", line 1288, in _send_request\n","    self.putrequest(method, url, **skips)\n","  File \"/usr/lib/python3.7/http/client.py\", line 1126, in putrequest\n","    self._output(self._encode_request(request))\n","  File \"/usr/lib/python3.7/http/client.py\", line 1206, in _encode_request\n","    return request.encode('ascii')\n","UnicodeEncodeError: 'ascii' codec can't encode characters in position 19-20: ordinal not in range(128)\n","127.0.0.1 - - [08/Jun/2021 08:32:33] \"\u001b[35m\u001b[1mGET /predict?url=http://ec2-52-79-44-86.ap-northeast-2.compute.amazonaws.com/upload/images/하이/1623141152.jpeg HTTP/1.1\u001b[0m\" 500 -\n","[2021-06-08 08:32:54,681] ERROR in app: Exception on /predict [GET]\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 2447, in wsgi_app\n","    response = self.full_dispatch_request()\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1952, in full_dispatch_request\n","    rv = self.handle_user_exception(e)\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1821, in handle_user_exception\n","    reraise(exc_type, exc_value, tb)\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/_compat.py\", line 39, in reraise\n","    raise value\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1950, in full_dispatch_request\n","    rv = self.dispatch_request()\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1936, in dispatch_request\n","    return self.view_functions[rule.endpoint](**req.view_args)\n","  File \"<ipython-input-15-6674ec62ed14>\", line 25, in predict_image\n","    img_path = save_and_get_img_path(url)\n","  File \"<ipython-input-6-b6e971d27ede>\", line 4, in save_and_get_img_path\n","    urllib.request.urlretrieve(url_path, filename=file_name)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 247, in urlretrieve\n","    with contextlib.closing(urlopen(url, data)) as fp:\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 222, in urlopen\n","    return opener.open(url, data, timeout)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 525, in open\n","    response = self._open(req, data)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 543, in _open\n","    '_open', req)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 503, in _call_chain\n","    result = func(*args)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 1378, in http_open\n","    return self.do_open(http.client.HTTPConnection, req)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 1350, in do_open\n","    encode_chunked=req.has_header('Transfer-encoding'))\n","  File \"/usr/lib/python3.7/http/client.py\", line 1277, in request\n","    self._send_request(method, url, body, headers, encode_chunked)\n","  File \"/usr/lib/python3.7/http/client.py\", line 1288, in _send_request\n","    self.putrequest(method, url, **skips)\n","  File \"/usr/lib/python3.7/http/client.py\", line 1126, in putrequest\n","    self._output(self._encode_request(request))\n","  File \"/usr/lib/python3.7/http/client.py\", line 1206, in _encode_request\n","    return request.encode('ascii')\n","UnicodeEncodeError: 'ascii' codec can't encode characters in position 19-20: ordinal not in range(128)\n","127.0.0.1 - - [08/Jun/2021 08:32:54] \"\u001b[35m\u001b[1mGET /predict?url=http://ec2-52-79-44-86.ap-northeast-2.compute.amazonaws.com/upload/images/하이/1623141174.jpeg HTTP/1.1\u001b[0m\" 500 -\n","[2021-06-08 08:33:11,571] ERROR in app: Exception on /predict [GET]\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 2447, in wsgi_app\n","    response = self.full_dispatch_request()\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1952, in full_dispatch_request\n","    rv = self.handle_user_exception(e)\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1821, in handle_user_exception\n","    reraise(exc_type, exc_value, tb)\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/_compat.py\", line 39, in reraise\n","    raise value\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1950, in full_dispatch_request\n","    rv = self.dispatch_request()\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1936, in dispatch_request\n","    return self.view_functions[rule.endpoint](**req.view_args)\n","  File \"<ipython-input-15-6674ec62ed14>\", line 26, in predict_image\n","    preprocess_image = prepare_image(img_path)\n","  File \"<ipython-input-7-0889dc54c20f>\", line 2, in prepare_image\n","    img = image.load_img(img_path, target_size=(img_width, img_height))\n","  File \"/usr/local/lib/python3.7/dist-packages/keras/preprocessing/image.py\", line 293, in load_img\n","    target_size=target_size, interpolation=interpolation)\n","  File \"/usr/local/lib/python3.7/dist-packages/keras_preprocessing/image/utils.py\", line 113, in load_img\n","    with open(path, 'rb') as f:\n","FileNotFoundError: [Errno 2] No such file or directory: ''\n","127.0.0.1 - - [08/Jun/2021 08:33:11] \"\u001b[35m\u001b[1mGET /predict?url=http://ec2-52-79-44-86.ap-northeast-2.compute.amazonaws.com/upload/images/ HTTP/1.1\u001b[0m\" 500 -\n","127.0.0.1 - - [08/Jun/2021 08:33:11] \"\u001b[33mGET /favicon.ico HTTP/1.1\u001b[0m\" 404 -\n","127.0.0.1 - - [08/Jun/2021 08:33:53] \"\u001b[37mGET /predict?url=http://ec2-52-79-44-86.ap-northeast-2.compute.amazonaws.com/upload/images/gkdl/1623141218.jpeg HTTP/1.1\u001b[0m\" 200 -\n","127.0.0.1 - - [08/Jun/2021 08:43:12] \"\u001b[37mPOST //video_list?key=AIzaSyCqEZTPQlS48rUKlRxe2sMvV1UXpK4F3ZM&&query=실내운동 HTTP/1.1\u001b[0m\" 200 -\n","127.0.0.1 - - [08/Jun/2021 08:44:20] \"\u001b[37mGET /predict?url=http://ec2-52-79-44-86.ap-northeast-2.compute.amazonaws.com/upload/images/gkdl/1623141859.jpeg HTTP/1.1\u001b[0m\" 200 -\n","127.0.0.1 - - [08/Jun/2021 08:44:30] \"\u001b[37mGET /predict?url=http://ec2-52-79-44-86.ap-northeast-2.compute.amazonaws.com/upload/images/gkdl/1623141868.jpeg HTTP/1.1\u001b[0m\" 200 -\n","127.0.0.1 - - [08/Jun/2021 08:52:41] \"\u001b[37mPOST //video_list?key=AIzaSyCqEZTPQlS48rUKlRxe2sMvV1UXpK4F3ZM&&query=실내운동 HTTP/1.1\u001b[0m\" 200 -\n","127.0.0.1 - - [08/Jun/2021 08:54:16] \"\u001b[37mPOST //video_list?key=AIzaSyCqEZTPQlS48rUKlRxe2sMvV1UXpK4F3ZM&&query=실내운동 HTTP/1.1\u001b[0m\" 200 -\n","127.0.0.1 - - [08/Jun/2021 08:54:45] \"\u001b[37mGET /predict?url=http://ec2-52-79-44-86.ap-northeast-2.compute.amazonaws.com/upload/images/kakaow/1623142484.jpeg HTTP/1.1\u001b[0m\" 200 -\n","[2021-06-08 08:55:56,228] ERROR in app: Exception on /predict [GET]\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 2447, in wsgi_app\n","    response = self.full_dispatch_request()\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1952, in full_dispatch_request\n","    rv = self.handle_user_exception(e)\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1821, in handle_user_exception\n","    reraise(exc_type, exc_value, tb)\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/_compat.py\", line 39, in reraise\n","    raise value\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1950, in full_dispatch_request\n","    rv = self.dispatch_request()\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1936, in dispatch_request\n","    return self.view_functions[rule.endpoint](**req.view_args)\n","  File \"<ipython-input-15-6674ec62ed14>\", line 25, in predict_image\n","    img_path = save_and_get_img_path(url)\n","  File \"<ipython-input-6-b6e971d27ede>\", line 4, in save_and_get_img_path\n","    urllib.request.urlretrieve(url_path, filename=file_name)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 247, in urlretrieve\n","    with contextlib.closing(urlopen(url, data)) as fp:\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 222, in urlopen\n","    return opener.open(url, data, timeout)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 525, in open\n","    response = self._open(req, data)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 543, in _open\n","    '_open', req)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 503, in _call_chain\n","    result = func(*args)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 1378, in http_open\n","    return self.do_open(http.client.HTTPConnection, req)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 1350, in do_open\n","    encode_chunked=req.has_header('Transfer-encoding'))\n","  File \"/usr/lib/python3.7/http/client.py\", line 1277, in request\n","    self._send_request(method, url, body, headers, encode_chunked)\n","  File \"/usr/lib/python3.7/http/client.py\", line 1288, in _send_request\n","    self.putrequest(method, url, **skips)\n","  File \"/usr/lib/python3.7/http/client.py\", line 1126, in putrequest\n","    self._output(self._encode_request(request))\n","  File \"/usr/lib/python3.7/http/client.py\", line 1206, in _encode_request\n","    return request.encode('ascii')\n","UnicodeEncodeError: 'ascii' codec can't encode characters in position 19-20: ordinal not in range(128)\n","127.0.0.1 - - [08/Jun/2021 08:55:56] \"\u001b[35m\u001b[1mGET /predict?url=http://ec2-52-79-44-86.ap-northeast-2.compute.amazonaws.com/upload/images/안녕/1623142556.jpeg HTTP/1.1\u001b[0m\" 500 -\n","[2021-06-08 08:55:58,903] ERROR in app: Exception on /predict [GET]\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 2447, in wsgi_app\n","    response = self.full_dispatch_request()\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1952, in full_dispatch_request\n","    rv = self.handle_user_exception(e)\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1821, in handle_user_exception\n","    reraise(exc_type, exc_value, tb)\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/_compat.py\", line 39, in reraise\n","    raise value\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1950, in full_dispatch_request\n","    rv = self.dispatch_request()\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1936, in dispatch_request\n","    return self.view_functions[rule.endpoint](**req.view_args)\n","  File \"<ipython-input-15-6674ec62ed14>\", line 25, in predict_image\n","    img_path = save_and_get_img_path(url)\n","  File \"<ipython-input-6-b6e971d27ede>\", line 4, in save_and_get_img_path\n","    urllib.request.urlretrieve(url_path, filename=file_name)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 247, in urlretrieve\n","    with contextlib.closing(urlopen(url, data)) as fp:\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 222, in urlopen\n","    return opener.open(url, data, timeout)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 525, in open\n","    response = self._open(req, data)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 543, in _open\n","    '_open', req)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 503, in _call_chain\n","    result = func(*args)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 1378, in http_open\n","    return self.do_open(http.client.HTTPConnection, req)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 1350, in do_open\n","    encode_chunked=req.has_header('Transfer-encoding'))\n","  File \"/usr/lib/python3.7/http/client.py\", line 1277, in request\n","    self._send_request(method, url, body, headers, encode_chunked)\n","  File \"/usr/lib/python3.7/http/client.py\", line 1288, in _send_request\n","    self.putrequest(method, url, **skips)\n","  File \"/usr/lib/python3.7/http/client.py\", line 1126, in putrequest\n","    self._output(self._encode_request(request))\n","  File \"/usr/lib/python3.7/http/client.py\", line 1206, in _encode_request\n","    return request.encode('ascii')\n","UnicodeEncodeError: 'ascii' codec can't encode characters in position 19-20: ordinal not in range(128)\n","127.0.0.1 - - [08/Jun/2021 08:55:58] \"\u001b[35m\u001b[1mGET /predict?url=http://ec2-52-79-44-86.ap-northeast-2.compute.amazonaws.com/upload/images/안녕/1623142558.jpeg HTTP/1.1\u001b[0m\" 500 -\n","[2021-06-08 08:56:01,636] ERROR in app: Exception on /predict [GET]\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 2447, in wsgi_app\n","    response = self.full_dispatch_request()\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1952, in full_dispatch_request\n","    rv = self.handle_user_exception(e)\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1821, in handle_user_exception\n","    reraise(exc_type, exc_value, tb)\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/_compat.py\", line 39, in reraise\n","    raise value\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1950, in full_dispatch_request\n","    rv = self.dispatch_request()\n","  File \"/usr/local/lib/python3.7/dist-packages/flask/app.py\", line 1936, in dispatch_request\n","    return self.view_functions[rule.endpoint](**req.view_args)\n","  File \"<ipython-input-15-6674ec62ed14>\", line 25, in predict_image\n","    img_path = save_and_get_img_path(url)\n","  File \"<ipython-input-6-b6e971d27ede>\", line 4, in save_and_get_img_path\n","    urllib.request.urlretrieve(url_path, filename=file_name)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 247, in urlretrieve\n","    with contextlib.closing(urlopen(url, data)) as fp:\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 222, in urlopen\n","    return opener.open(url, data, timeout)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 525, in open\n","    response = self._open(req, data)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 543, in _open\n","    '_open', req)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 503, in _call_chain\n","    result = func(*args)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 1378, in http_open\n","    return self.do_open(http.client.HTTPConnection, req)\n","  File \"/usr/lib/python3.7/urllib/request.py\", line 1350, in do_open\n","    encode_chunked=req.has_header('Transfer-encoding'))\n","  File \"/usr/lib/python3.7/http/client.py\", line 1277, in request\n","    self._send_request(method, url, body, headers, encode_chunked)\n","  File \"/usr/lib/python3.7/http/client.py\", line 1288, in _send_request\n","    self.putrequest(method, url, **skips)\n","  File \"/usr/lib/python3.7/http/client.py\", line 1126, in putrequest\n","    self._output(self._encode_request(request))\n","  File \"/usr/lib/python3.7/http/client.py\", line 1206, in _encode_request\n","    return request.encode('ascii')\n","UnicodeEncodeError: 'ascii' codec can't encode characters in position 19-20: ordinal not in range(128)\n","127.0.0.1 - - [08/Jun/2021 08:56:01] \"\u001b[35m\u001b[1mGET /predict?url=http://ec2-52-79-44-86.ap-northeast-2.compute.amazonaws.com/upload/images/안녕/1623142561.jpeg HTTP/1.1\u001b[0m\" 500 -\n","127.0.0.1 - - [08/Jun/2021 09:06:57] \"\u001b[37mPOST //video_list?key=AIzaSyCqEZTPQlS48rUKlRxe2sMvV1UXpK4F3ZM&&query=실내운동 HTTP/1.1\u001b[0m\" 200 -\n","127.0.0.1 - - [08/Jun/2021 09:07:19] \"\u001b[37mPOST //video_list?key=AIzaSyCqEZTPQlS48rUKlRxe2sMvV1UXpK4F3ZM&&query=실내운동 HTTP/1.1\u001b[0m\" 200 -\n","127.0.0.1 - - [08/Jun/2021 09:07:47] \"\u001b[37mGET /predict?url=http://ec2-52-79-44-86.ap-northeast-2.compute.amazonaws.com/upload/images/hihello/1623143266.jpeg HTTP/1.1\u001b[0m\" 200 -\n","127.0.0.1 - - [08/Jun/2021 09:08:37] \"\u001b[37mGET /predict?url=http://ec2-52-79-44-86.ap-northeast-2.compute.amazonaws.com/upload/images/dkeke/1623143316.jpeg HTTP/1.1\u001b[0m\" 200 -\n","127.0.0.1 - - [08/Jun/2021 09:15:01] \"\u001b[37mPOST //video_list?key=AIzaSyCqEZTPQlS48rUKlRxe2sMvV1UXpK4F3ZM&&query=실내운동 HTTP/1.1\u001b[0m\" 200 -\n","127.0.0.1 - - [08/Jun/2021 09:16:27] \"\u001b[37mGET /predict?url=http://ec2-52-79-44-86.ap-northeast-2.compute.amazonaws.com/upload/images/great/1623143785.jpeg HTTP/1.1\u001b[0m\" 200 -\n"],"name":"stderr"}]}]}